torch/csrc/utils/throughput_benchmark.cpp

```diff
@@ -0,0 +1,129 @@
+#include <torch/csrc/utils/throughput_benchmark.h>
+
+#include <torch/csrc/jit/pybind_utils.h>
+#include <torch/csrc/utils/auto_gil.h>
+
+namespace torch {
+namespace throughput_benchmark {
+
+void ThroughputBenchmark::addInput(py::args args, py::kwargs kwargs) {
+  CHECK(script_module_.initialized() ^ module_.initialized());
+  if (script_module_.initialized()) {
+    script_module_.addInput(std::move(args), std::move(kwargs));
+  } else {
+    CHECK(module_.initialized());
+    module_.addInput(std::move(args), std::move(kwargs));
+  }
+}
+
+py::object ThroughputBenchmark::runOnce(py::args&& args, py::kwargs&& kwargs)  {
+  CHECK(script_module_.initialized() ^ module_.initialized());
+  if (script_module_.initialized()) {
+    c10::IValue result;
+    {
+      AutoNoGIL no_gil_guard;
+      result = script_module_.runOnce(std::move(args), std::move(kwargs));
+    }
+    return jit::toPyObject(std::move(result));
+  } else {
+    CHECK(module_.initialized());
+    return module_.runOnce(std::move(args), std::move(kwargs));
+  }
+}
+
+ThroughputBenchmark::ThroughputBenchmark(
+    std::shared_ptr<jit::script::Module> script_module)
+    : script_module_(std::move(script_module)) {}
+
+ThroughputBenchmark::ThroughputBenchmark(
+    py::object module)
+    : module_(std::move(module)) {}
+
+BenchmarkExecutionStats ThroughputBenchmark::benchmark(
+    const BenchmarkConfig& config) const {
+  CHECK(script_module_.initialized() ^ module_.initialized());
+  // Main benchmark thread doesn't hold the GIL after scheduling worker threads
+  // But for now we don't release it as we will be implicitly manipulating with
+  // py::object ref. counts in the case of nn.Module benchmarking.
+  if (script_module_.initialized()) {
+    return script_module_.benchmark(config);
+  } else {
+    CHECK(module_.initialized());
+    TORCH_WARN("Starting benchmark on an nn.Module. This can be slow due "
+    "to Python GIL.For proper inference simulation you might want to switch to "
+    "a ScriptModule instead");
+    return module_.benchmark(config);
+  }
```

Comment:
Since everything else is templated, why isn't this class? Every single method has an `if (script_modue_.initialized())` so it's basically trying to do two completely different things.

Reply:
This is so we can have one object  in the python bindings, this class hides the if statements from users